### ISSUE TITLE
[css-display][css-ui] Should outline apply to elements with 'display: contents'?

### ISSUE THREAD

**Issue by fantasai** [w3c/csswg-drafts#3323]
[François Remy wrote](https://github.com/w3c/csswg-drafts/issues/2632#issuecomment-393062411):

> I  think outline should work on display contents elements. Just like it  works if an online gets split into lines or bidi or even regions or by a  block-in-inline. I think the element should be focusable and  tab-focusable too, but only if it is not replaced or a shadow host or  none of its descendants in the light tree produce a box. The elements in  cases noted above behave as if they were display none and should  therefore not be focusable. Besides outline and a few other properties  like pointer-event, properties should have no effect on the elements  that do generate boxes for their descendants, besides being inherited by  default.
>
>I think a case can be made here.
> 
>> Outlines may be non-rectangular. For example, if the element is broken across several lines, the outline should be an outline or minimum set of outlines that encloses all the element’s boxes.
>
>> The parts of the outline are not required to be rectangular. To the extent that the outline follows the border edge, it should follow the border-radius curve.
>
>> The position of the outline may be affected by descendant boxes.
>
> Now to the question "does the spec say you should have an outline in that case" I'd say probably no. Whether it should say so is another question. You will have to answer these questions anyway when working on exposing these elements to assistive technologies. If these elements are there for semantic purposes, they must be exposed, and you will somehow need to synthetize something for them.

**Comment by fantasai** [comment]
@frivoal Yes, it's normatively defined in css-display-3. https://www.w3.org/TR/css-display-3/#the-display-properties

**Comment by dbaron** [comment]
While the spec does say (as quoted above):

> The position of the outline may be affected by descendant boxes.

Gecko has been having to move away from doing this (or at least seriously consider doing so) for web compatibility, since other browsers don't do it (at least in many cases).

So I'd be pretty hesitant to add new behavior that depends on it, unless other browsers are committing to making outlines enclose descendant boxes.

**Comment by FremyCompany** [comment]
I'd like to note that we have had multiple discussions on this topic both in other issues and outside of github in the past few weeks, and I have come to revise my opinion here a bit. 

I think it wouldn't make no sense to decide that `display:contents` do not render outlines, since they do not have boxes on which the outline drawing request can be stored. In the end, what I think really matters is that when a `display:contents` link gets keyboard-focused, an outline is drawn around the boxes it's associated with, but that can be achieved by "manually giving a default `auto` outline to the boxed descendants of an un-boxed focused element" rather than having an element without a box draw an multipart outline. 

Both options are possible, really, and I'd be fine with both of these ways of doing things, but my understanding of the current consensus from people invovled in this issue is that, implementation-wise, not drawing an outline for a `display:contents` element makes more a lot more sense, and I'm inclined to believe it.

**Comment by SelenIT** [comment]
I agree with @FremyCompany. Absence of focus outline for the element with `display:contents` itself would be completely OK as long as the element remains focusable _and_ keyboard-accessible, and its `:focus` state can be used to style its descendants and pseudo-elements.

**Comment by dbaron** [comment]
Is our resolution clear about what should happen in case (2) here?

```html
<div style="display: contents; outline: 1px dotted">
    <div>(1) This div clearly has an outline.</div>
    <div style="display:contents">
        <div>(2) Does this div have an outline?</div>
    </div>
</div>
```

**Comment by SelenIT** [comment]
My understanding is that it implies recursive propagation (as mentioned in the discussion), so I'd say that the resulting outline should span both rendered divs.

**Comment by annevk** [comment]
Reading the discussion I'm not sure that was considered and I also recall @MatsPalmgren pretty explicitly objecting to this in a Bugzilla bug.

I'm rather worried that this case-by-case decision-making regarding which aspects of a box matter and which don't for `display:contents` will lead to a rather confused design.

**Comment by SelenIT** [comment]
I'm afraid that _any_ design of the feature like `display:contents` (aptly dubbed `display: schrodinger-cat;` in the discussion:) couldn't completely avoid any confusion. But the proposed solution looks significantly less confusing for users and authors than the existing implementations, and thus better conforming to the Priorities of Constituencies principle.

> case-by-case decision-making regarding which aspects of a box matter and which don't

As far as I understand, it's not the change in "which _aspects of the box_ matter", it's "which _boxes_ matter" that is changing. When the element clearly _has a visible and interactive box tree,_ making the closest visible boxes in that tree matter instead of _the only absent box_ in it makes much sense for me.

**Comment by annevk** [comment]
It does not make sense to me, since [above](https://github.com/w3c/csswg-drafts/issues/3323#issuecomment-499529450) you're arguing for enclosing two child boxes with a new wrapper box of some kind that has the outline...

**Comment by SelenIT** [comment]
I don't argue for that. I only said that, according to my understanding of the CSSWG resolution, both boxes should have an outline around them. I have no preference how exactly this compound outline should be implemented, it's up to vendors.

The only thing that I really argue for is that element that has _any_ visible and interactive _box tree_ should be focusable and keyboard-accessible by default.

**Comment by dbaron** [comment]
So given the more recent discussions in https://github.com/whatwg/html/pull/9425 and https://github.com/mozilla/standards-positions/issues/772 and elsewhere, I'm coming back to this issue.

I think the idea of propagating the outline to children (possibly recursively) mostly makes sense, although it may or may not be quite difficult to implement.  I think there are a few issues we'd need to sort out at a conceptual level, though:

* Is it a "merged" outline or separate rectangles around each child?  (Chrome does "merged" outlines for multiple lines of links, and I think this sort of merging was part of the original intent of `outline`.  I'm not sure about the current state in other browsers.)

* How do we handle children that are text nodes?  Text nodes normally don't draw outlines.

If we do a "merged" outline I think we could associate it with the first relevant child (for z-ordering, filtering, etc. purposes).  Or maybe the last one if outlines aren't (in some engines, at least) implemented at a special position in the z-order.

I think the difficult parts from an implementation perspective probably relate to getting invalidation right when things change (such as changes to the size of the children or addition or removal of children).
